src/transcript/keccak.rs

```diff
@@ -0,0 +1,92 @@
+use std::marker::PhantomData;
+use tiny_keccak::{Keccak, Hasher};
+use ark_ec::CurveGroup;
+use ark_ff::{BigInteger, PrimeField};
+
+use crate::transcript::Transcript;
+
+/// KecccakTranscript implements the Transcript trait using the Keccak hash
+pub struct KeccakTranscript<C: CurveGroup> {
+    sponge: Keccak,
+    phantom: PhantomData<C>,
+}
+
+#[derive(Debug)]
+pub struct KeccakConfig {}
+
+impl<C: CurveGroup> Transcript<C> for KeccakTranscript<C> {
+    type TranscriptConfig = KeccakConfig;
+    fn new(config: &Self::TranscriptConfig) -> Self {
+        let _ = config;
+        let sponge = Keccak::v256();
+        Self {
+            sponge,
+            phantom: PhantomData,
+        }
+    }
+
+    fn absorb(&mut self, v: &C::ScalarField) {
+        self.sponge.update(&(v.into_bigint().to_bytes_le()));
+    }
+    fn absorb_vec(&mut self, v: &[C::ScalarField]) {
+        for _v in v {
+            self.sponge.update(&(_v.into_bigint().to_bytes_le()));
+        }
+    }
+    fn absorb_point(&mut self, p: &C) {
+        let mut serialized = vec![];
+        p.serialize_compressed(&mut serialized).unwrap();
+        self.sponge.update(&(serialized))
+    }
+    fn get_challenge(&mut self) -> C::ScalarField {
+        let mut output = [0u8; 32];
+        self.sponge.clone().finalize(&mut output);
+        self.sponge.update(&[output[0]]);
```

Comment:
Should we re-initialize the sponge here? I saw other impls just drops the inputs after absorbing them.

See:
- https://github.com/AztecProtocol/barretenberg/blob/9de42a37415f8a3f1bd7de3ec460b1b299dd15fa/sol/src/ultra/BaseUltraVerifier.sol
- https://github.com/matter-labs/solidity_plonk_verifier/blob/f6eea6bde11321d3104b20d100e9b6051102a2a2/bellman_vk_codegen/template.sol#L264

Reply:
@han0110 Thanks for pointing out!

You mean we don't need to call `absorb()` (or `update()`) after squeeze() here, right?

I just checked, and it seems that since `finalize()` on tiny-keccak/SHA3 does not update the original sponge, thus we no need to call absorb() afterwards as you mentioned.

I have pushed a fix and new tests which can make sure that. ✅ https://github.com/privacy-scaling-explorations/folding-schemes/pull/27/commits/0685d375fb2a213173e5f4ada84bc5d0276cc506

Reply:
hmm not really, I mean we should re-initialize `self.sponge` after we've called `finalize`, otherwise next time we call `get_challenge` it's equal to hashing all the previous thing again, for evm verifier it seems inefficient.

so it would look like:

```diff
         let mut output = [0u8; 32];
-        self.sponge.clone().finalize(&mut output);
+        std::mem::replace(&mut self.sponge, Keccak::v256()).finalize(&mut output);
         self.sponge.update(&[output[0]]);
```